### PR TITLE
User balance

### PR DESCRIPTION
**Pull request by devton**

### CODE REVIEW

File: src/c/admin-balance-transfer-item-detail.js [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+import _ from 'underscore';
+import postgrest from 'mithril-postgrest';
+import h from '../h';
+import userVM from '../vms/user-vm';
+import models from '../models';
+import adminUserBalanceTransactionsList from './admin-user-balance-transactions-list';
+
+const adminBalanceTransferItemDetail = {
+    controller(args) {
+        const userBankAccount = m.prop(undefined),
```
Comment (vicnicius) [inline src/c/admin-balance-transfer-item-detail.js]: I think we should replace `undefined` for `null`.

File: src/c/admin-balance-transfer-item-detail.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+                  m.request({
+                      method: 'PUT',
+                      url: `/admin/balance_transfers/${args.item.id}`,
+                      data: {
+                          balance_transfer: {
+                              admin_notes: fields.admin_notes()
+                          }
+                      },
+                      config: h.setCsrfToken
+                  }).then((data) => {
```
Comment (vicnicius) [inline src/c/admin-balance-transfer-item-detail.js]: No need for the data parameter here. Let's try to always add the catch method to our requests also.

File: src/c/admin-balance-transfer-item-detail.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+                      loadingNotes(false);
+                  });
+              };
+
+        if(!_.isUndefined(metaBank)) {
+            if(metaBank.conta) {
+                transitionBankAccount({
+                    account: metaBank.conta,
+                    account_digit: metaBank.conta_dv,
+                    account_type: undefined,
```
Comment (vicnicius) [inline src/c/admin-balance-transfer-item-detail.js]: same thing about replacing `undefined` for `null`.

File: src/c/admin-balance-transfer-item-detail.js [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+                    agency: metaBank.agencia,
+                    agency_digit: metaBank.agencia_dv,
+                    bank_code: metaBank.bank_code,
+                    bank_name: undefined,
```
Comment (vicnicius) [inline src/c/admin-balance-transfer-item-detail.js]: `undefined` for `null`

File: src/c/admin-balance-transfer-item-detail.js [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+                    owner_document: metaBank.document_number,
+                    owner_name: metaBank.legal_name
+                });
+            } else {
+                transitionBankAccount(metaBank);
+            }
+        }
+
+        userVM.getUserBankAccount(args.item.user_id).then(_.compose(userBankAccount, _.first));
```
Comment (vicnicius) [inline src/c/admin-balance-transfer-item-detail.js]: add a catch method

File: src/c/admin-balance-transfer-item-detail.js [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+                        ` ${bankAccount.agency} - ${bankAccount.agency_digit ? bankAccount.agency_digit : ''}`,m('br'),
+                        m('span.fontweight-semibold', "Conta:"),
+                        ` ${bankAccount.account} - ${bankAccount.account_digit ? bankAccount.account_digit : ''}`,m('br'),
+                        m('span.fontweight-semibold', 'Nome:'),
+                        bankAccount.owner_name, m('br'),
+                        m('span.fontweight-semibold', 'CPF:'),
+                        bankAccount.owner_document
+                    ])
+                    ] : h.loader()),
+                    (ctrl.loadingNotes() ? h.loader() : m('', [
```
Comment (vicnicius) [inline src/c/admin-balance-transfer-item-detail.js]: empty tag string here. in this case it is fine to simply return the array if a wrapper is not needed. : )

File: src/c/admin-balance-transfer-item.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+                m('.fontcolor-secondary.fontsize-smallest',
+                  item.user_email),
+                m('.fontcolor-secondary.fontsize-smallest',
+                  `USER_ID: ${item.user_id}`)
+             ]),
+            m('.w-col.w-col-2', [
+                m('span.fontsize-small', `R$ ${h.formatNumber(item.amount, 2, 3)}`)
+            ]),
+            m('.w-col.w-col-2.w-hidden-small.w-hidden-tiny', [
+                m('', [
```
Comment (vicnicius) [inline src/c/admin-balance-transfer-item.js]: empty m(''). we should either just remove this wrapper and return the array or make it an actual tag.

File: src/c/admin-user-balance-transactions-list.js [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    controller(args) {
+        const userBalance = m.prop({}),
+              transactionsListVM = postgrest.paginationVM(
+                  models.balanceTransaction,
+                  'created_at.desc',
+                  { Prefer: 'count=exact'}
+              );
+
+        models.balanceTransaction.pageSize(2);
+        userVM.getUserBalance(args.user_id).then(_.compose(userBalance, _.first));
```
Comment (vicnicius) [inline src/c/admin-user-balance-transactions-list.js]: add catch method

File: src/c/filter-dropdown.js [hunk 9]
```diff
@@ -9,7 +9,7 @@ const filterDropdown = {
               (args.custom_label ? m.component(args.custom_label[0], args.custom_label[1]) : args.label)),
             m.component(dropdown, {
                 id: args.index,
-                onchange: args.onchange,
+                onchange: _.isFunction(args.onchange) ? args.onchange : () => {},
```
Comment (vicnicius) [inline src/c/filter-dropdown.js]: just a note: I would rather write noops as `Function.prototype`

File: src/c/menu-profile.js [hunk 10]
```diff
@@ -21,25 +24,36 @@ const menuProfile = {
 
         userVM.fetchUser(user_id, true, userDetails);
 
+        userIdVM.user_id(user_id);
+        models.balance.getRowWithToken(userIdVM.parameters()).then((result) => {
```
Comment (vicnicius) [inline src/c/menu-profile.js]: We should add a catch method and provide an error message, if possible.

File: src/c/project-successful-onboard.js [hunk 11]
```diff
@@ -22,114 +21,58 @@ const I18nScope = _.partial(h.i18nScope, 'projects.successful_onboard');
...
+                      tax_link: {
+                          action: 'onclick',
+                          actionSource: () => {
+                              showTaxModal.toggle();
+                              m.redraw();
+                          }
+                      }
+                  };
+
+                  if (isInitialized) return;
```
Comment (vicnicius) [inline src/c/project-successful-onboard.js]: This could be before the toRedraw definition.

File: src/c/user-balance-request-modal-content.js [hunk 12]
```diff
@@ -12,97 +12,209 @@ import _ from 'underscore';
...
+                      if (parsedErrors) {
+                          parsedErrors.resetFieldErrors();
+                      }
+                      parsedErrors = userSettingsVM.mapRailsErrors(err.errors_json);
+                      loading(false);
+                      m.redraw();
+                  });
+              },
+              requestFund = () => {
+                  requestLoader.load().then((data) => {
```
Comment (vicnicius) [inline src/c/user-balance-request-modal-content.js]: Missing catch error and error feedback.

File: src/c/user-balance-request-modal-content.js [hunk 13]
```diff
@@ -12,97 +12,209 @@ import _ from 'underscore';
...
         };
     },
     view(ctrl, args) {
-        const balance = args.balance;
+        const balance = args.balance,
+              fields = ctrl.fields,
+              user = args.user;
 
-        return (ctrl.loadBankA() ? h.loader() : m('div', _.map(ctrl.bankAccounts(), item => [
+        return (m('div', [
```
Comment (vicnicius) [inline src/c/user-balance-request-modal-content.js]: remove extra parenthesis wrapping return.

File: src/c/user-bank-form.js [hunk 14]
```diff
@@ -0,0 +1,302 @@
...
+                  id: '127',
+                  code: '399',
+                  name: 'HSBC Bank Brasil S.A. - Banco Múltiplo'
+              }, {
+                  id: '23',
+                  code: '237',
+                  name: 'Banco Bradesco S.A.'
+              }];
+
+        userVM.getUserBankAccount(user.id).then((data) => {
```
Comment (vicnicius) [inline src/c/user-bank-form.js]: Add catch method and display error feedback

File: src/c/user-bank-form.js [hunk 15]
```diff
@@ -0,0 +1,302 @@
...
+                fields.agency(bankAccount().agency);
+                fields.agency_digit(bankAccount().agency_digit);
+                fields.bank_id(bankAccount().bank_id);
+                fields.bank_account_type(bankAccount().account_type);
+                args.bankCode(bankAccount().bank_id);
+            } else {
+                fields.bank_account_type('conta_corrente');
+            }
+        });
+        banksLoader.load().then(banks);
```
Comment (vicnicius) [inline src/c/user-bank-form.js]: missing catch.

File: src/root/admin-balance-tranfers.js [hunk 16]
```diff
@@ -0,0 +1,330 @@
...
+                      return i.id == item.id;
+                  });
+                  selectedItemsIDs(newIDs);
+                  if(_.isEmpty(newIDs)) {
+                      selectedAny(false);
+                  }
+              },
+              submit = () => {
+                  error(false);
+                  listVM.firstPage(filterVM.parameters()).then(null, (serverError) => {
```
Comment (vicnicius) [inline src/root/admin-balance-tranfers.js]: missing catch handler.

File: src/root/admin-balance-tranfers.js [hunk 17]
```diff
@@ -0,0 +1,330 @@
...
+                  m.request({
+                      method: 'POST',
+                      url: '/admin/balance_transfers/batch_manual',
+                      data: {
+                          transfer_ids: _.uniq(_.map(selectedItemsIDs(), (s) => {
+                              return s.id;
+                          }))
+                      },
+                      config: h.setCsrfToken
+                  }).then((data) => {
```
Comment (vicnicius) [inline src/root/admin-balance-tranfers.js]: missing catch handler

File: src/root/admin-balance-tranfers.js [hunk 18]
```diff
@@ -0,0 +1,330 @@
...
+                  m.request({
+                      method: 'POST',
+                      url: '/admin/balance_transfers/batch_approve',
+                      data: {
+                          transfer_ids: _.uniq(_.map(selectedItemsIDs(), (s) => {
+                              return s.id;
+                          }))
+                      },
+                      config: h.setCsrfToken
+                  }).then((data) => {
```
Comment (vicnicius) [inline src/root/admin-balance-tranfers.js]: missing catch handler

File: src/root/admin-balance-tranfers.js [hunk 19]
```diff
@@ -0,0 +1,330 @@
...
+                  m.request({
+                      method: 'POST',
+                      url: '/admin/balance_transfers/batch_reject',
+                      data: {
+                          transfer_ids: _.uniq(_.map(selectedItemsIDs(), (s) => {
+                              return s.id;
+                          }))
+                      },
+                      config: h.setCsrfToken
+                  }).then((data) => {
```
Comment (vicnicius) [inline src/root/admin-balance-tranfers.js]: missing catch handler.

File: src/root/admin-contributions.js [hunk 20]
```diff
@@ -138,7 +138,7 @@ const adminContributions = {
     },
 
     view(ctrl) {
-        return [
+        return m('', [
```
Comment (vicnicius) [inline src/root/admin-contributions.js]: When we need a wrapper, we can add at least an ID or something to help identify the node inside the DOM.

File: src/c/admin-balance-transfer-item-detail.js (hunk 6 above)
Comment (devton) [inline src/c/admin-balance-transfer-item-detail.js]: wrapper is needed because is already inside an array

**Pull request merged**